AppController/djinn.rb

```diff
@@ -5427,6 +5451,107 @@ def scale_deployment
+        "so will not be scaling down")
+      return
+    end
+
+    # Also, don't scale down if we just scaled up or down.
+    if Time.now.to_i - @last_scaling_time < (SCALEUP_THRESHOLD * 
+        SCALE_TIME_MULTIPLIER * DUTY_CYCLE)
+      Djinn.log_info("Not scaling down right now, as we recently scaled " +
+        "up or down.")
+      return
+    end
+    
+    # Look through an array of autoscaled nodes and check if any of the 
+    # machines are not running any AppServers and need to be downscaled.
+    get_autoscaled_nodes.reverse_each { |node|
+      break if num_scaled_down == max_scale_down_capacity
+      
+      hosted_apps = []
+      @apps_loaded.each { |app_name|
+        @app_info_map[app_name]['appengine'].each { |location|
+          host, port = location.split(":")
+          if host == node.private_ip
+            hosted_apps << "#{app_name}:#{port}"
+          end
+        }
+      }
+      
+      unless hosted_apps.empty?
+        Djinn.log_debug("The node #{node.private_ip} has these AppServers " +
+        "running: #{hosted_apps}")
+        next
+      end
+
+      # Right now, only the autoscaled machines are started with just the 
+      # appengine role, so we check specifically for that during downscaling
+      # to make sure we only downscale the new machines added.
+      node_to_remove = nil
+      if node.jobs == ['appengine']
+        Djinn.log_info("Removing node #{node}")
+        node_to_remove = node
+      end
+      
+      num_terminated = terminate_node_from_deployment(node_to_remove)
+      num_scaled_down += num_terminated
+    }
+    @last_scaling_time = Time.now.to_i
+  end
+
+  # Removes the specified node from the deployment and terminates
+  # the instance from the cloud.
+  #
+  # Args:
+  #   node_to_remove: A node instance, to be terminated and removed 
+  #     from this deployment.
+  def terminate_node_from_deployment(node_to_remove)
+    if node_to_remove.nil?
+      Djinn.log_warn("Tried to scale down but couldn't find a node to remove.")
+      return 0
+    end
+
+    remove_node_from_local_and_zookeeper(node_to_remove.private_ip)
+
+    to_remove = {}
+    @app_info_map.each { |_, info|
+      next if info['appengine'].nil?
+
+      info['appengine'].each { |location|
+        host, port = location.split(":")
+        if host == node_to_remove.private_ip
+          to_remove[app] = [] if to_remove[app].nil?
+          to_remove[app] << location
+        end
+      }
+    }
+    to_remove.each { |app, locations|
+        locations.each { |location|
+          @app_info_map[app]['appengine'].delete(location)
+        }
+    }
+
+    imc = InfrastructureManagerClient.new(@@secret)
+    begin
+      imc.terminate_instances(@options, node_to_remove.instance_id)
+    rescue FailedNodeException
+      Djinn.log_warn("Failed to call terminate_instances")
+      return 0
+    end
+
+    @last_scaling_time = Time.now.to_i
```

Comment:
Yes, good point. I will put it in a lock and a thread because the terminate can take a long enough time.